### PR TITLE
sla: Added packages that implement the support api

### PR DESCRIPTION
**Pull request by mattyw**
package that implement a client who can request support macaroons. Still early stages

### CODE REVIEW

File: api/support/api.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+	"github.com/juju/errors"
+	"github.com/juju/romulus/wireformat/support"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+	"gopkg.in/macaroon.v1"
+)
+
+var DefaultURL = "https://api.jujucharms.com/omnibus/v2"
+
+// SupportAuthClient defines the interface available to clients of the support api.
```
Comment (alesstimec) [inline api/support/api.go]: rename to SLAAuthClient

File: api/support/api.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+type SupportAuthClient interface {
+	// Authorize returns the support macaroon for the specified model
+	Authorize(modelUUID, supportLevel string) (*macaroon.Macaroon, error)
```
Comment (alesstimec) [inline api/support/api.go]: please use:

type SLA string

const (
	Unsupported SLA = "unsupported"
	Essential   SLA = "essential"
	Standard    SLA = "standard"
	Advanced    SLA = "advanced"
)

File: api/support/api.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+var _ SupportAuthClient = (*client)(nil)
+
+type httpClient interface {
+	// Do sends the given HTTP request and returns its response.
+	Do(*http.Request) (*http.Response, error)
```
Comment (alesstimec) [inline api/support/api.go]: we don't need Do atm.. we only do a POST, which uses DoWithBody.. min interfaces, etc..

File: api/support/api.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+// BaseURL sets the base url for the api client.
+func BaseURL(url string) func(h *client) error {
+	return func(h *client) error {
+		h.baseURL = url
+		return nil
+	}
+}
+
+// NewSupportAuthClient returns a new support authorization client.
+func NewSupportAuthClient(options ...ClientOption) (SupportAuthClient, error) {
```
Comment (alesstimec) [inline api/support/api.go]: why do we have NewSupportAuthClient and NewClient? perhaps the NewClient does not need to be exported?

File: api/support/api.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return c, nil
+}
+
+// Authorize obtains a support authorization macaroon.
+func (c *client) Authorize(modelUUID, supportLevel string) (*macaroon.Macaroon, error) {
+	u, err := url.Parse(c.baseURL + "/support/authorize")
```
Comment (alesstimec) [inline api/support/api.go]: /sla/authorize

File: api/support/api.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+	if response.StatusCode != http.StatusOK {
+		body, err := ioutil.ReadAll(response.Body)
+		if err == nil {
+			return nil, errors.Errorf("failed to authorize support level: received http response: %v - code %q", string(body), http.StatusText(response.StatusCode))
+		}
+		return nil, errors.Errorf("failed to authorize support level: http response is %q", http.StatusText(response.StatusCode))
+	}
+
+	var m *macaroon.Macaroon
```
Comment (alesstimec) [inline api/support/api.go]: var m macaroon.Macaroon

File: api/support/api.go (hunk 2 above)
Comment (mattyw) [inline api/support/api.go]: @alesstimec I think that's a decision for the server to make. Not the client

File: api/support/api.go (hunk 1 above)
Comment (alesstimec) [inline api/support/api.go]: hmm.. support.SupportAuthClient.. stuttering.. can we have just support.AuthClient?

File: wireformat/sla/entities.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// The sla package implements wireformats for the sla service.
+package sla
+
+// SupportRequest defines the json used to post to sla service.
```
Comment (cmars) [inline wireformat/sla/entities.go]: SLARequest

File: api/sla/api.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	response, err := c.client.DoWithBody(req, bytes.NewReader(buff.Bytes()))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer discardClose(response)
+
+	if response.StatusCode != http.StatusOK {
+		body, err := ioutil.ReadAll(response.Body)
+		if err == nil {
+			return nil, errors.Errorf("failed to authorize support level: received http response: %v - code %q", string(body), http.StatusText(response.StatusCode))
```
Comment (cmars) [inline api/sla/api.go]: failed to authorize SLA

File: api/sla/api.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+		}
+		return nil, errors.Errorf("failed to authorize support level: http response is %q", http.StatusText(response.StatusCode))
```
Comment (cmars) [inline api/sla/api.go]: failed to authorize SLA

File: api/sla/api.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+	return m, nil
+}
+
+// discardClose reads any remaining data from the response body and closes it.
+func discardClose(response *http.Response) {
+	if response == nil || response.Body == nil {
+		return
+	}
+	io.Copy(ioutil.Discard, response.Body)
```
Comment (cmars) [inline api/sla/api.go]: log.Debugf the error here and next line?

File: api/sla/api.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return c, nil
+}
+
+// Authorize obtains a sla authorization macaroon.
+func (c *client) Authorize(modelUUID, supportLevel string) (*macaroon.Macaroon, error) {
```
Comment (mattyw) [inline api/sla/api.go]: need to add budget

File: api/sla/api.go (hunk 10 above)
Comment (mattyw) [inline api/sla/api.go]: We shouldn't need to as we're just discarding anything that wasn't read

**Comment by mattyw** [comment]
$$merge$$

**Pull request merged**